src/app.rs

```diff
@@ -185,55 +181,53 @@ impl App {
         let event_source = EventSource::new()?;
         let rx_events = event_source.receiver();
 
-        // create the initial state
-        if let Some(bs) = BrowserState::new(
-            con.launch_args.root.clone(),
-            con.launch_args.tree_options.clone(),
-            &screen,
-            &TaskLifetime::unlimited(),
-        )? {
-            self.push(Box::new(bs));
-        } else {
-            unreachable!();
-        }
+        self.push(Box::new(
+            BrowserState::new(
+                con.launch_args.root.clone(),
+                con.launch_args.tree_options.clone(),
+                &screen,
+                &TaskLifetime::unlimited(),
+            )?
+            .expect("Failed to create BrowserState"),
+        ));
 
         let mut cmd = Command::new();
 
         // if some commands were passed to the application
         //  we execute them before even starting listening for events
         if let Some(unparsed_commands) = &con.launch_args.commands {
-            let commands = parse_command_sequence(unparsed_commands, con)?;
-            for arg_cmd in &commands {
-                cmd = (*arg_cmd).clone();
-                cmd = self.apply_command(writer, cmd, &mut screen, con)?;
-                self.do_pending_tasks(writer, &cmd, &mut screen, con, TaskLifetime::unlimited())?;
+            let lifetime = TaskLifetime::unlimited();
+
+            for arg_cmd in parse_command_sequence(unparsed_commands, con)? {
+                cmd = self.apply_command(writer, arg_cmd, &mut screen, con)?;
+                self.do_pending_tasks(writer, &cmd, &mut screen, con, lifetime.clone())?;
                 if self.quitting {
                     return self.end(writer);
                 }
             }
         }
 
-        self.mut_state().display(writer, &screen, con)?;
-        self.state().write_status(writer, &cmd, &screen, con)?;
-        self.state().write_flags(writer, &mut screen, con)?;
+        let state = self.mut_state();
+        state.display(writer, &screen, con)?;
+        state.write_status(writer, &cmd, &screen, con)?;
+        state.write_flags(writer, &mut screen, con)?;
+
         screen.input_field.display_on(writer)?;
 
-        loop {
-            let tl = TaskLifetime::new(event_source.shared_event_count());
-            if !self.quitting {
```

Comment:
Correctness note: the only way that self.quitting can be true is through a call to `apply_command`, which means that moving it to the end of the loop is acceptable here